src/batch/src/executor/delete.rs

```diff
@@ -96,33 +100,49 @@ impl DeleteExecutor {
         let data_types = self.child.schema().data_types();
         let mut builder = DataChunkBuilder::new(data_types, 1024);
 
+        let table_dml_handle = self
+            .dml_manager
+            .table_dml_handle(self.table_id, self.table_version_id)?;
+        let mut write_handle = table_dml_handle.write_handle(self.txn_id)?;
+
         let mut notifiers = Vec::new();
 
+        notifiers.push(write_handle.begin()?);
+
         // Transform the data chunk to a stream chunk, then write to the source.
-        let write_chunk = |chunk: DataChunk| async {
+        let write_txn_data = |chunk: DataChunk| {
             let cap = chunk.capacity();
             let stream_chunk = StreamChunk::from_parts(vec![Op::Delete; cap], chunk);
 
-            self.dml_manager
-                .write_chunk(self.table_id, self.table_version_id, stream_chunk)
-                .await
+            #[cfg(debug_assertions)]
+            table_dml_handle.check_chunk_schema(&stream_chunk);
+
+            write_handle.write_chunk(stream_chunk)
         };
 
         #[for_await]
         for data_chunk in self.child.execute() {
-            let data_chunk = data_chunk?;
+            let data_chunk = match data_chunk {
+                Ok(data_chunk) => data_chunk,
+                Err(err) => {
+                    write_handle.rollback()?;
```

Comment:
By returning from this stream, the `write_handle` will be automatically dropped by RAII so I guess we don't have to explicitly call `rollback` here.